core/src/main/java/org/apache/accumulo/core/client/admin/TableOperations.java

```diff
@@ -1014,39 +1014,15 @@ default TimeType getTimeType(String tableName) throws TableNotFoundException {
   }
 
   /**
-   * Initiates setting a table to an on-demand state, but does not wait for action to complete
+   * Sets the hosting goal for a range of Tablets in the specified table.
    *
-   * @param tableName the table to set to an on-demand state
-   * @throws AccumuloException when there is a general accumulo error
-   * @throws AccumuloSecurityException when the user does not have the proper permissions
-   * @since 3.1.0
-   */
-  void onDemand(String tableName)
-      throws AccumuloSecurityException, AccumuloException, TableNotFoundException;
-
-  /**
-   * Initiates setting a table to an on-demand state, optionally waits for action to complete
-   *
-   * @param tableName the table to set to an on-demand state
-   * @param wait if true, then will not return until table state is set to an on-demand state
-   * @throws AccumuloException when there is a general accumulo error
-   * @throws AccumuloSecurityException when the user does not have the proper permissions
-   * @since 3.1.0
-   */
-  void onDemand(String tableName, boolean wait)
-      throws AccumuloSecurityException, AccumuloException, TableNotFoundException;
-
-  /**
-   * Check if a table is on-demand through its current goal state only. Could run into issues if the
-   * current state of the table is in between states. If you require a specific state, call
-   * <code>onDemand(tableName, true)</code>, this will wait until the table reaches the desired
-   * state before proceeding.
-   *
-   * @param tableName the table to check if in an on-demand state
-   * @throws AccumuloException when there is a general accumulo error
-   * @return true if table's goal state is online
-   * @since 3.1.0
+   * @param tableName table name
+   * @param range tablet range
+   * @param goal hosting goal (one of {@literal ALWAYS, DEFAULT, NEVER})
```

Comment:
Is DEFAULT like ondemand?  I do not have a sense of what it means from the name.  ALWAYS and NEVER are very clear to me.

Reply:
> When a tablet is unloaded, it's hosting_goal column value is set to default.

Seems like this should only change when a user changes it.  If a user specifies they want a tablet to always be hosted and the balancer moves the tablet then its goal state should still be ALWAYS after the move.

> So there are 3 user states and 4 internal states. We need an internal state for on-demand and don't host (default) and on-demand and host (ondemand)

Maybe we could have three user facing states `NEVER,ONDEMAND,ALWAYS` and two columns in the metadata table. A `hosting_goal` column that stores the user facing goal states.  A `hosting_needed` column that is used to indicate a tablet with an ondemand goal state needs hosting.  The `hosting_goal`  is only modified by user operations.  The `hosting_needed` is modified by the system as needed.

> A non-existent hosting_goal column, or one with an empty Value, is treated like default.

 We could also treat any tablet with an unspecified hosting goal as an error if we set them on upgrade and in initialization.  Setting it on upgrade can make the code simpler as it can be written with the expectation that its always set.

Reply:
If we expect the hosting goal to always be set, then modifying the upgrade code to set it could be a follow on PR.

Reply:
I did the following in d579d7b:

1. Allowed the user to set hosting goal to ALWAYS, NEVER, or ONDEMAND. Removed the DEFAULT option.
2. Stored the user set hosting goal in the tablet metadata in the column `hosting:goal`
3. Default hosting goal value of user tables is ONDEMAND
4. When client requests a Tablet to be hosted, the column `hosting:requested` will be added to the tablet metadata.
5. When the TabletServer unloads an ondemand tablet, it will remove the `hosting:requested` column from the tablet metadata.